pallets/marketplace/src/lib.rs

```diff
@@ -768,7 +771,54 @@ pub mod pallet {
                     <StoredJobStatus<T>>::remove(&job_id.0, &job_id.1);
                     <StoredJobRegistration<T>>::remove(&job_id.0, &job_id.1);
                 }
-                _ => Err(Error::<T>::JobRegistrationUnmodifiable)?,
+                JobStatus::Matched => {
+                    T::MarketplaceHooks::finalize_job(job_id, T::RewardManager::refund(job_id))?;
+
+                    // Remove matching data
+
+                    for (p, _) in <AssignedProcessors<T>>::iter_prefix(&job_id) {
+                        <StoredMatches<T>>::remove(&p, &job_id);
+                    }
+
+                    let _ =
+                        <AssignedProcessors<T>>::clear_prefix(&job_id, T::MaxSlots::get(), None);
+                    <StoredJobStatus<T>>::remove(&job_id.0, &job_id.1);
+                    <StoredJobRegistration<T>>::remove(&job_id.0, &job_id.1);
+                }
+                JobStatus::Assigned(_) => {
+                    // Get the job requirements
+                    let registration = <StoredJobRegistration<T>>::get(&job_id.0, &job_id.1)
+                        .ok_or(pallet_acurast::Error::<T>::JobRegistrationNotFound)?;
+                    let extra: <T as Config>::RegistrationExtra = registration.extra.clone().into();
+                    let requirements: JobRequirementsFor<T> = extra.into();
+
+                    // Compute the reward amount to be payed to each assigned processor
+                    let remaining_reward = Self::reserved(job_id);
+                    let reward_per_processor = remaining_reward
+                        .checked_div(&(requirements.slots.into()))
+                        .ok_or(Error::<T>::UnexpectedCheckedCalculation)?;
+
+                    // Pay reward to the processor and clear matching data
+                    for (processor, _) in <AssignedProcessors<T>>::iter_prefix(&job_id) {
+                        match T::ManagerProvider::manager_of(&processor) {
+                            Ok(manager) => T::RewardManager::pay_reward(
+                                &job_id,
+                                reward_per_processor,
+                                &manager,
+                            ),
+                            Err(err_result) => Err(err_result.into()),
+                        }?;
+                        <StoredMatches<T>>::remove(&processor, &job_id);
+                    }
+
+                    // The job creator will only receive the amount that could not be divided equaly between the processors
```

Comment:
Comment incorrect:  job creator also receives the reward of processors that did not (yet) acknowledge the assignment to them. A test to show this would also be nice